### PR TITLE
Add automation for Restart command for K3s 

### PR DESCRIPTION
**Pull request by est-suse**
#### Proposed Changes ####

Automation to test restart server and agents from k3s

#### Types of Changes ####

Test 
#### Verification ####
Run the test 

#### Testing ####
N/A

#### Linked Issues ####

https://github.com/rancher/qa-tasks/issues/588

#### User-Facing Change ####

```release-note
NONE
```

### CODE REVIEW

File: tests/e2e/validatecluster/validatecluster_test.go [hunk 1]
```diff
@@ -197,6 +198,41 @@ var _ = Describe("Verify Create", Ordered, func() {
 			}, "420s", "10s").Should(Succeed())
 		})
 
+		It("Verifies Restart K3s", func() {
```
Comment (brandond) [inline tests/e2e/validatecluster/validatecluster_test.go:201]: This claims to be just verifying that the service is restarted; what does the daemonset deployment have to do with verifying the restart?

File: tests/e2e/validatecluster/validatecluster_test.go [hunk 2]
```diff
@@ -169,6 +169,7 @@ var _ = Describe("Verify Create", Ordered, func() {
 
 			for _, nodeName := range serverNodeNames {
 				ip, _ := e2e.FetchNodeExternalIP(nodeName)
+				println(ip)
```
Comment (brandond) [inline tests/e2e/validatecluster/validatecluster_test.go:172]: is this desired, or just for debugging?

File: tests/e2e/testutils.go [hunk 3]
```diff
@@ -399,6 +399,17 @@ func RestartCluster(nodeNames []string) error {
 	return nil
 }
 
+// RestartCluster restarts the k3s service on each agents given
+func RestartAgents(nodeNames []string) error {
+	for _, nodeName := range nodeNames {
+		cmd := "sudo systemctl restart k3s-Agent"
```
Comment (brandond) [inline tests/e2e/testutils.go:405]:
```suggestion
		cmd := "sudo systemctl restart k3s-agent"
```

**Review by brandond — CHANGES_REQUESTED** [review]

File: tests/e2e/testutils.go [hunk 4]
```diff
@@ -399,6 +399,17 @@ func RestartCluster(nodeNames []string) error {
 	return nil
 }
 
+// RestartCluster restarts the k3s service on each agents given
```
Comment (briandowns) [inline tests/e2e/testutils.go:402]: Update doc comment to reflect change.

File: tests/e2e/testutils.go (hunk 3 above)
Comment (briandowns) [inline tests/e2e/testutils.go:405]: This should be made a const

**Comment by dereknola** [comment]
This is partially being addressed by https://github.com/k3s-io/k3s/pull/6890 with the `RestartCluster` function.

File: tests/e2e/validatecluster/validatecluster_test.go [hunk 5]
```diff
@@ -189,14 +184,40 @@ var _ = Describe("Verify Create", Ordered, func() {
...
 
+		It("Verifies Restart K3s", func() {
+			_, err := e2e.DeployWorkload("daemonset.yaml", kubeConfigFile, *hardened)
+			Expect(err).NotTo(HaveOccurred(), "Daemonset manifest not deployed")
+			nodes, _ := e2e.ParseNodes(kubeConfigFile, false)
+
+			Eventually(func(g Gomega) {
+				pods, _ := e2e.ParsePods(kubeConfigFile, false)
+				count := e2e.CountOfStringInSlice("test-daemonset", pods)
+				g.Expect(len(nodes)).Should((Equal(count)), "Daemonset pod count does not match node count")
```
Comment (rancher-max) [inline tests/e2e/validatecluster/validatecluster_test.go:199]: Can we also add an expectation here to ensure that the daemonset pods are Running? This so far just validates that they are deployed to the nodes, but they could be Crashlooping

File: tests/e2e/validatecluster/validatecluster_test.go (hunk 1 above)
Comment (est-suse) [reply tests/e2e/validatecluster/validatecluster_test.go:191]: The daemonset is useful for making sure that non-system pods continue to work after a restart.

File: tests/e2e/validatecluster/validatecluster_test.go [hunk 6]
```diff
@@ -189,14 +184,40 @@ var _ = Describe("Verify Create", Ordered, func() {
...
+			Eventually(func(g Gomega) {
+				nodes, err := e2e.ParseNodes(kubeConfigFile, false)
+				g.Expect(err).NotTo(HaveOccurred())
+				for _, node := range nodes {
+					g.Expect(node.Status).Should(Equal("Ready"))
+				}
+				cmd := "kubectl get pods -o=name -l k8s-app=test-daemonset --field-selector=status.phase=Running --kubeconfig=" + kubeConfigFile
+				res, err := e2e.RunCommand(cmd)
+				Expect(err).NotTo(HaveOccurred())
+				g.Expect(res).Should(ContainSubstring("test-daemonset"), "daemonset pods not Running")
```
Comment (rancher-max) [inline tests/e2e/validatecluster/validatecluster_test.go:215]: Can we change this logic to use the same functions as pre-restart?

Also, add another expect here to ensure the daemonset pods are all Running after the restart.

**Review by rancher-max — CHANGES_REQUESTED** [review]

File: tests/e2e/validatecluster/validatecluster_test.go (hunk 6 above)
Comment (est-suse) [reply tests/e2e/validatecluster/validatecluster_test.go:215]: Thanks but What about this block of code that is executed after the restart:
```
cmd := "kubectl get pods -o=name -l k8s-app=test-daemonset --field-selector=status.phase=Running --kubeconfig=" + kubeConfigFile
				res, err := e2e.RunCommand(cmd)
				Expect(err).NotTo(HaveOccurred())
				g.Expect(res).Should(ContainSubstring("test-daemonset"), "daemonset pods not Running")
```

it is to ensure the daemonset pods are all Running after the restart.

Comment (rancher-max) [reply tests/e2e/validatecluster/validatecluster_test.go:215]: That code gets the name of all pods in the `test-daemonset` app that are currently Running. 

However, it's possible that one of them is running and another one is Crashlooping, but in this case you would not see that. Also, it's possible that it's state is technically "Running", but it's stuck at "0/1 Ready". Using the same functions as earlier in the code, we can get these results and check that they are what we expect them to be

File: tests/e2e/validatecluster/validatecluster_test.go [hunk 7]
```diff
@@ -189,14 +184,51 @@ var _ = Describe("Verify Create", Ordered, func() {
...
+			nodes, _ := e2e.ParseNodes(kubeConfigFile, false)
+			Eventually(func(g Gomega) {
+				pods, _ := e2e.ParsePods(kubeConfigFile, false)
+				count := e2e.CountOfStringInSlice("test-daemonset", pods)
+				g.Expect(len(nodes)).Should((Equal(count)), "Daemonset pod count does not match node count")
+				podsRunning := 0
+				for _, pod := range pods {
+					if strings.Contains(pod.Name, "test-daemonset") && pod.Status == "Running" {
+						podsRunning++
+					}
```
Comment (rancher-max) [inline tests/e2e/validatecluster/validatecluster_test.go:203]: One more thing to add in here that I missed the first time around: `Ready`. Should be x/x (1/1 or 2/2 or whatever, just not something like 1/2 or 0/1). We should check **both here before the restart and after the restart** that all of the pods in this daemonset are also Ready.

**Review by rancher-max — CHANGES_REQUESTED** [review]
Two more small changes but otherwise LGTM!

**Review by rancher-max — APPROVED** [review]
LGTM

File: tests/e2e/testutils.go [hunk 8]
```diff
@@ -391,7 +400,18 @@ func ParsePods(kubeConfig string, print bool) ([]Pod, error) {
 // RestartCluster restarts the k3s service on each node given
 func RestartCluster(nodeNames []string) error {
```
Comment (brandond) [inline tests/e2e/testutils.go:401]: Should this be RestartServer, since it only works on server nodes?

**Review by brandond — APPROVED** [review]
LGTM, other than the nit about fixing the RestartCluster function name

File: tests/e2e/testutils.go (hunk 8 above)
Comment (est-suse) [reply tests/e2e/testutils.go:401]: Thanks Brad, but I did not create the RestartCluster, if I change the name it'll require to change it in any other file that is using the func. I did add the docstring.

**Review by rancher-max — APPROVED** [review]

**Review by dereknola — APPROVED** [review]

**Pull request merged**